src/potdata/sampler.py

```diff
@@ -206,23 +274,134 @@ def sample(self, data: list[Structure]) -> list[Structure]:
 
         return selected
 
-    def plot(self):
+    def sample_points(self, reduced_vectors, eps=0.00019, min_samples=4, noisy_ratio=1, reachable_ratio=1, core_ratio=0.3):
+        db = DBSCAN(eps=eps, min_samples=min_samples).fit(reduced_vectors)
+        labels = db.labels_
+
+        core_samples_mask = np.zeros_like(labels, dtype=bool)
+        core_samples_mask[db.core_sample_indices_] = True
+
+        # Sample noisy points
+        noisy_points = reduced_vectors[labels == -1]
+        sample_size_noisy = int(noisy_ratio * len(noisy_points))
+        sampled_noisy_points = random.sample(list(noisy_points), sample_size_noisy)
+
+        # Sample reachable points
+        reachable_points = reduced_vectors[np.logical_and(labels != -1, ~core_samples_mask)]
+        sample_size_reachable = int(reachable_ratio * len(reachable_points))
+        sampled_reachable_points = random.sample(list(reachable_points), sample_size_reachable)
+
+        # Sample core points
+        core_points = reduced_vectors[core_samples_mask]
+        sample_size_core = int(core_ratio * len(core_points))
+        sampled_core_points = random.sample(list(core_points), sample_size_core)
+
+        return sampled_noisy_points, sampled_reachable_points, sampled_core_points, db
+    
+    def plot_clusters_with_sampling(self, eps=0.00019, min_samples=4, noisy_ratio=1, reachable_ratio=1, core_ratio=0.3):
+        
         """Function to plot the results of the selection.
 
         This can be called after the `sample` method to visualize the results.
         """
+        
+        soap_vectors = self._get_soap_vectors(data)
 
-    def _get_soap_vectors(self, data: list[Structure]) -> list[np.ndarray]:
-        """Convert structures to SOAP vectors."""
+        if self.post_soap_selection is not None:
+            soap_vectors = self.post_soap_selection(data, soap_vectors)
 
-    def _dim_reduction(self, vectors: list[np.ndarray], dim: int):
-        """Perform dimension reduction on the SOAP vectors."""
+        vectors = np.concatenate(soap_vectors, axis=1)
 
-    def _cluster(self, data: list[Structure]):
-        """Perform DBSCAN."""
+        if self.pca_dim is not None:
+            vectors = self._dim_reduction(vectors, self.pca_dim)
+
+        db = DBSCAN(eps=eps, min_samples=min_samples).fit(vectors)
```

Comment:
You should not call any DBSCAN here. In the previous steps, you've already obtained the info such as the labels. Here you just plot them.

Reply:
Fixed